### PR TITLE
Annotations for public API

### PR DESCRIPTION
**Pull request by KmolYuan**
#523

The generic type syntax `Class[T]` is supported by `__class_getitem__` magic method but generate `Class[T] <=> Class` during runtime. (since Python 3.7)

The subclass of `_Decoratable` and `VisitorBase` are supported generic typing. `Transformer` can be typed as:

```python
class T(Transformer[float]):
    ...

t = T()
f = t.transform(tree)  # type: float
```

Some methods not mentioned on documentation are not recorded.

Requires Python 3.8.

### CODE REVIEW

File: lark/tree.pyi [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+from .lexer import Token
+
+
+class Tree:
+
+    data: str
+    children: List[Union[str, Tree]]
+    meta: Token
+
+    def __init__(self, data: str, children: List[Tree], meta: Optional[Token] = None):
```
Comment (erezsh) [inline lark/tree.pyi]: `children` should accept the same type as `Tree.children`

`meta` is not a `Token`, it's `tree.Meta`.

File: lark/tree.pyi [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    def find_data(self, data: str) -> Iterator[Tree]:
+        ...
+
+    def iter_subtrees(self) -> Iterator[Tree]:
+        ...
+
+    def iter_subtrees_topdown(self) -> Iterator[Tree]:
+        ...
+
+    def __eq__(self, other: object) -> bool:
```
Comment (erezsh) [inline lark/tree.pyi]: Why is this necessary? No one calls `__eq__` directly. Perhaps there's a reason I'm not aware of?

File: lark/exceptions.pyi [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class LexError(LarkError):
+    pass
+
+
+class UnexpectedInput(LarkError):
+    pos_in_stream: int
+
+    def get_context(self, text: str, span: int = 40):
```
Comment (erezsh) [inline lark/exceptions.pyi]: Isn't it a bad idea to write the default here? It may change in the future.

Or is it unused?

File: lark/lark.pyi [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    def __init__(
+        self,
+        grammar: Union[str, IO[str]],
+        *,
+        start: _Start = ...,
+        parser: Literal["earley", "lalr", "cyk"] = ...,
+        lexer: Optional[Lexer] = ...,
+        transformer: Optional[Transformer] = ...,
+        postlex: Optional[Literal["standard", "contextual"]] = ...,
```
Comment (erezsh) [inline lark/lark.pyi]: `postlex` is a callable object (a lexer proxy)

However
```python
  lexer: Optional[Literal["standard", "contextual", "dynamic", "dynamic_complete"]]
```

File: lark/exceptions.pyi [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# -*- coding: utf-8 -*-
+
+from typing import Dict, Iterable, Callable
+from .tree import Tree
+
+
+class LarkError(Exception):
```
Comment (erezsh) [inline lark/exceptions.pyi]: Why are the empty classes required? Does mypy gain any information from this?

File: lark/lexer.pyi [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        ...
+
+
+class TerminalDef:
+    name: str
+    pattern: Pattern
+    priority: int
+
+
+class Token(str):
```
Comment (erezsh) [inline lark/lexer.pyi]: Missing `Token.value`

**Comment by KmolYuan** [comment]
> Also, is it possible to put all the pyi files in their own folder? I would like the keep the code namespace clean and for code only.

Yes, I can pack them into a dummy module called "`lark-stubs`".
It can work with both development and users.

Sorry about missing some members in stubs, because there are not mentioned in docs.
I'll add them as soon as you instruct.

File: lark/exceptions.pyi (hunk 5 above)
Comment (KmolYuan) [reply lark/exceptions.pyi]: MyPy will check inheritance of the classes. Subclass instance is allow to passed as a superclass instance. ([ref](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#class-types))
Is this exception public? If some one want to catch all Lark exceptions.

File: lark/exceptions.pyi (hunk 3 above)
Comment (KmolYuan) [reply lark/exceptions.pyi]: Default value can show as `span: int = ...` if it might changed.
Or just hide it if it's a developer option.

File: lark/lexer.pyi (hunk 6 above)
Comment (KmolYuan) [reply lark/lexer.pyi]: Docs not mentioned. I thought it was used to represent itself in `str(self)`.

File: lark/tree.pyi (hunk 2 above)
Comment (KmolYuan) [reply lark/tree.pyi]: Well, in this case it's unnecessary. Or represent a override method?

File: lark/tree.pyi (hunk 1 above)
Comment (KmolYuan) [reply lark/tree.pyi]: Sorry, because I saw [doc](https://lark-parser.readthedocs.io/en/latest/classes/#tree) mentioned it has some attributes not listed.
Should I add in `tree.Meta` (as dynamic attributes) or implement as a [`typing.Protocol`](https://docs.python.org/3/library/typing.html#typing.Protocol)?

Comment (erezsh) [reply lark/tree.pyi]: I'm not sure how Protocol will help. I think it's fine to write `Meta`'s attributes as if they are there, since they usually will be (unless empty is True)

File: lark/tree.pyi (hunk 2 above)
Comment (erezsh) [reply lark/tree.pyi]: What do you mean?

File: lark/exceptions.pyi (hunk 3 above)
Comment (erezsh) [reply lark/exceptions.pyi]: Can't mypy take it from the actual function? Or is there a way to use a CONST from the code (i.e. `DEFAULT_CONTEXT_SPAN`, so there will be one source of truth?

File: lark/lark.pyi (hunk 4 above)
Comment (KmolYuan) [reply lark/lark.pyi]: What is `postlex` looks like?
I found an `Indenter` instance was used in there.
It's not exposed yet in my work.

File: lark/exceptions.pyi (hunk 5 above)
Comment (erezsh) [reply lark/exceptions.pyi]: `LarkError` should be public. So you can either fix that in your patch, or I'll do it in a different patch.

The same is true for most exceptions (most of them are already public)

File: lark/lexer.pyi (hunk 6 above)
Comment (erezsh) [reply lark/lexer.pyi]: It's better to look at the code.

**Comment by erezsh** [comment]
Great, sounds good. `stubs` is the idiomatic word? Because `type-annotations` sounds more informative.

**Comment by KmolYuan** [comment]
@erezsh Yes, a package contains only `pyi` files are named `*-stubs`.
It will be installed with original package and can be found by MyPy.

File: lark/lark.pyi (hunk 4 above)
Comment (erezsh) [reply lark/lark.pyi]: postlex is a Protocol with the method `process(self, stream: Iterator[Token]) -> Iterator[Token])`

File: lark/exceptions.pyi (hunk 3 above)
Comment (KmolYuan) [reply lark/exceptions.pyi]: The actual value is not important in static type checking.
It at most makes the IDE display its preset value.

File: lark/exceptions.pyi (hunk 5 above)
Comment (KmolYuan) [reply lark/exceptions.pyi]: I'll patch it in `lark/__init__.py`.

File: lark/tree.pyi (hunk 2 above)
Comment (KmolYuan) [reply lark/tree.pyi]: Since superclass `object` already have it, unless it's a abstract method, I will remove it.

File: lark/exceptions.pyi (hunk 3 above)
Comment (erezsh) [reply lark/exceptions.pyi]: My point is that if a value will be displayed, it has to be the correct one, and change if the code changes. Otherwise, it's better to not have a value.

Comment (KmolYuan) [reply lark/exceptions.pyi]: I think the default value is just to remind the user that this is an optional parameter of default behavior.
Regardless the implementation, they should keep refer to docstring or documentation if they are not truly understand the function.

Comment (erezsh) [reply lark/exceptions.pyi]: A quick search revealed the correct way to write it:

    for arguments with a default, use ... instead of the actual default;

https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#conventions

File: lark-stubs/exceptions.pyi [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# -*- coding: utf-8 -*-
+
+from typing import Dict, Iterable, Callable
+from .tree import Tree
+
+
+class LarkError(Exception):
```
Comment (erezsh) [inline lark-stubs/exceptions.pyi]: So these empty exceptions can be removed now, right?

File: lark-stubs/lexer.pyi [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    mres: _MRes
+
+    def __init__(self, mres: _MRes):
+        ...
+
+    def __call__(self, t: Token) -> Token:
+        ...
+
+
+class CallChain:
```
Comment (erezsh) [inline lark-stubs/lexer.pyi]: Why did you add all these internal classes? I don't think 99% of the users will ever even see them..

Comment (KmolYuan) [reply lark-stubs/lexer.pyi]: Oh, That's the note when I'm tried to deriving member types in some methods, same as `_Lex`.
I can remove them.

File: lark-stubs/exceptions.pyi (hunk 7 above)
Comment (KmolYuan) [reply lark-stubs/exceptions.pyi]: The public exception should be listed to avoid import problem in MyPy, because `pyi` will replace `py` module when doing type check.

Comment (erezsh) [reply lark-stubs/exceptions.pyi]: Why would there be an import problem though?

File: lark-stubs/lexer.pyi (hunk 8 above)
Comment (erezsh) [reply lark-stubs/lexer.pyi]: The thing is that some of these classes might change in the future, because they aren't part of the API.

How does it help to have these? (For example with _Lex?)

File: lark-stubs/exceptions.pyi (hunk 7 above)
Comment (KmolYuan) [reply lark-stubs/exceptions.pyi]: For users checking their code. MyPy will check the annotated dependencies either.

File: lark-stubs/lexer.pyi (hunk 8 above)
Comment (KmolYuan) [reply lark-stubs/lexer.pyi]: I left it behind when assuming the types of some members, because they are generated from some non-public methods.

Comment (KmolYuan) [reply lark-stubs/lexer.pyi]: Like `CallChain` and `UnlessCallback` can be simplified as `Callable` but I don't known at first time.

File: lark-stubs/exceptions.pyi (hunk 7 above)
Comment (KmolYuan) [reply lark-stubs/exceptions.pyi]: Would it be better to identify the public name if you add a `__all__` list? ([PEP8](https://www.python.org/dev/peps/pep-0008/#module-level-dunder-names))

Comment (erezsh) [reply lark-stubs/exceptions.pyi]: If it's necessary and will allow to remove these re-definitions, that I'll consider it.

Is the an import problem currently? I don't understand why..

Comment (KmolYuan) [reply lark-stubs/exceptions.pyi]: Do you mean that empty inheritance like "LarkError" should be removed? If it is a public name, it can't, and its existence is necessary. If not, other subclasses can inherit from `Exception` directly.

Comment (erezsh) [reply lark-stubs/exceptions.pyi]: My question is why do exceptions need to be included at all? There is no additional information presented in them, and they are already accessible by importing exceptions.py?

Comment (KmolYuan) [reply lark-stubs/exceptions.pyi]: No, pyi will replace py files if exist.

<https://mypy.readthedocs.io/en/stable/stubs.html#creating-a-stub>

> If a directory contains both a .py and a .pyi file for the same module, the .pyi file takes precedence.

**Comment by erezsh** [comment]
Thanks!
